Cryptlib/SysCall/BaseMemAllocation.c

```diff
@@ -19,20 +19,63 @@ WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
 // -- Memory-Allocation Routines --
 //
 
+/* Prefix allocations with the allocation's total size.
+ *
+ * This is used to implement `realloc`, which needs to know the
+ * allocation's current size so that the data can be copied over to the
+ * new allocation.
+ */
+struct AllocWithSize {
+  /* Size of the whole allocation, including this `size` field.
+   *
+   * Allocations returned by AllocatePool are always eight-byte aligned
+   * (according to the UEFI spec), so use an eight-byte `size` field to
+   * make the `data` field also eight-byte aligned. */
+  UINT64 size;
+
+  /* The beginning of the allocation returned to the caller of malloc.
+   *
+   * The pointer to this data is what will then get passed into realloc
+   * and free; those functions use ptr_to_alloc_with_size to get the
+   * pointer to the underlying pool allocation. */
+  UINT8 data[0];
+};
+
+/* Convert a `ptr` (as returned by malloc) to an AllocWithSize by
+ * subtracting eight bytes. */
+static struct AllocWithSize *ptr_to_alloc_with_size (void *ptr) {
+  UINT8 *cptr = (UINT8*) ptr;
+  return (struct AllocWithSize*) (cptr - sizeof(struct AllocWithSize));
+}
+
 /* Allocates memory blocks */
 void *malloc (size_t size)
 {
-  return AllocatePool ((UINTN) size);
+  UINTN alloc_size = (UINTN) (size + sizeof(struct AllocWithSize));
+
+  struct AllocWithSize *alloc = AllocatePool (alloc_size);
+  if (alloc) {
+    alloc->size = alloc_size;
+    return alloc->data;
+  } else {
+    return NULL;
+  }

```

Comment:
It looks like github lost a review comment of mine here.

Here I pointed out that this should be formatted as follows -- no `else` after `return` please:

~~~
  if (alloc) {
    alloc->size = alloc_size;
    return alloc->data;
  }
  return NULL;
~~~